src/redux/api/templateApi.ts

```diff
@@ -0,0 +1,66 @@
+import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
+import { environment } from "@/constants/environment";
+import { ITemplate, ITemplateDataDb } from "@/interfaces/template";
+import _ from "lodash";
+
+export const templateApi = createApi({
+  reducerPath: "templates",
+  baseQuery: fetchBaseQuery({
+    baseUrl: environment.apiBaseUrl,
+  }),
+  tagTypes: ["template"],
+  endpoints: (builder) => ({
+    getTemplates: builder.query<Omit<ITemplateDataDb, "__v" | "isDeleted">[], string>({
+      query: (userId) => `/templates?user_id=${userId}`,
+      transformResponse: (result: ITemplateDataDb[], _meta, _arg) => {
+        return result.map((item) => _.omit(item, ["__v", "isDeleted"]));
+      },
+      providesTags: (result, _err, _arg) =>
+        result
+          ? [...result.map(({ _id }) => ({ type: "template" as const, id: _id })), "template"]
+          : ["template"],
+    }),
+
+    getTemplateById: builder.query<Omit<ITemplateDataDb, "__v" | "isDeleted">, string>({
+      query: (templateId) => `templates/${templateId}`,
+      transformResponse: (result: ITemplateDataDb, _meta, _arg) => {
+        return _.omit(result, ["__v", "isDeleted"]);
+      },
+      providesTags: (result, _err, _arg) =>
+        result ? [{ type: "template", id: result._id }] : ["template"],
+    }),
+
+    addTemplate: builder.mutation<ITemplateDataDb, Omit<ITemplate, "_id">>({
+      query: (newTemplate) => ({
+        url: "templates",
+        method: "POST",
+        body: newTemplate,
+      }),
+      invalidatesTags: ["template"],
+    }),
+
+    deleteTemplate: builder.mutation<boolean, string>({
+      query: (templateId) => ({
+        url: `/templates/${templateId}`,
+        method: "DELETE",
+      }),
+      invalidatesTags: (_result, _err, arg) => [{ type: "template", id: arg }],
+    }),
+
+    updateTemplate: builder.mutation<ITemplateDataDb, Partial<ITemplate> & Pick<ITemplate, "_id">>({
+      query: (templateid) => ({
+        url: `/templates/${templateid}`,
+        method: "PUT",
+      }),
+      invalidatesTags: (_result, _err, arg) => [{ type: "template", id: arg._id }],
```

Comment:
In my view, it is better to add {type:"template", id:"LIST"} after { type: "template", id: arg._id }. When a specified template was updated, then the cache of the template with that specified id was affected, but the "LIST" of all templates in the template community was not updated. Users need to reload the template community page by themselves if they want to see the updated template which they updated just a few seconds ago. Add {type:"template", id:"LIST"} here, as well as at the providesTags at getTemplates endpoint can solve this issue.

Reply:
> In my view, it is better to add {type:"template", id:"LIST"} after { type: "template", id: arg._id }. When a specified template was updated, then the cache of the template with that specified id was affected, but the "LIST" of all templates in the template community was not updated. Users need to reload the template community page by themselves if they want to see the updated template which they updated just a few seconds ago. Add {type:"template", id:"LIST"} here, as well as at the providesTags at getTemplates endpoint can solve this issue.

I defined  [{ type: "template", id: _id_1}, {type: "template", id: _id_2}, {type: "template", id: _id_3}, ... ... ] rather than {type: "template" id: "LIST"} in useQuery endpoints. It has the same effect as what you said. For example, when {type: "template", id: 123456} is invalid in the mutation endpoints, it will do the re-fetch action in useQuery which contain the tag {type: "template", id: 123456}.

Reply:
👍